File: src/Player.cpp
/*=============================================================================
 * TarotClub - Player.cpp
 *=============================================================================
 * Basic modelization of a tarot player
 *=============================================================================
 * TarotClub ( http://www.tarotclub.fr ) - This file is part of TarotClub
 * Copyright (C) 2003-2999 - Anthony Rabine
 * anthony@tarotclub.fr
 *
 * TarotClub is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * TarotClub is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with TarotClub.  If not, see <http://www.gnu.org/licenses/>.
 *
 *=============================================================================
 */

#include "Player.h"
#include "Protocol.h"

/*****************************************************************************/
Player::Player()
    : mUuid(Protocol::INVALID_UID)
    , mAck(false)
{

}
/*****************************************************************************/
std::uint32_t Player::GetUuid() const
{
    return mUuid;
}
/*****************************************************************************/
void Player::SetUuid(std::uint32_t value)
{
    mUuid = value;
}
/*****************************************************************************/
Deck Player::AutoDiscard(const Deck &dog, std::uint8_t nbPlayers)
{
    Deck discard;

    // We add all the dog cards to the player's deck
    Append(dog);

    // We're looking valid discard cards to put in the discard
    for (Deck::ConstIterator iter = Begin(); iter != End(); ++iter)
    {
        Card c = (*iter);
        if ((c.GetSuit() != Card::TRUMPS) && (c.GetValue() != 14U))
        {
            discard.Append(c);

            if (discard.Size() == Tarot::NumberOfDogCards(nbPlayers))
            {
                // enough cards!
                break;
            }
        }
    }

    return discard;
}
/*****************************************************************************/
/**
 * @brief Player::canPlayCard
 *
 * Test if the card cVerif can be played depending of the already played cards
 * and the cards in the player's hand.
 *
 * @param mainDeck
 * @param cVerif
 * @param gameCounter
 * @param nbPlayers
 * @return true if the card can be played
 */
bool Player::CanPlayCard(const Card &card, Deck &trick)
{
    std::uint8_t   suit; // required suit
    bool ret = false;
    PlayerStats stats;

    // Check if the player has the card in hand
    if (!HasCard(card))
    {
        return false;
    }

    // The player is the first of the trick, he can play any card
    if (trick.Size() == 0)
    {
        return true;
    }

    // Simple use case, the excuse can always be played
    if (card.IsFool())
    {
        return true;
    }

    // We retreive the requested suit by looking at the first card played
    Card c = *(trick.Begin());

    if (c.IsFool())
    {
        // The first card is a Excuse...
        if (trick.Size() == 1)
        {
            // ...the player can play everything he wants
            return true;
        }
        // The requested suit is the second card
        c = *(++trick.Begin());
    }
    suit = c.GetSuit();

    // Some indications about previous played cards
    for (Deck::ConstIterator it = trick.Begin(); it != trick.End(); ++it)
    {
        c = *(it);
        if (c.GetSuit() == Card::TRUMPS)
        {
            stats.previousTrump = true;
            if (c.GetValue() > stats.maxPreviousTrump)
            {
                stats.maxPreviousTrump = c.GetValue();
            }
        }
    }

    // Some indications on the player cards in hand
    for (Deck::ConstIterator it = Begin(); it != End(); ++it)
    {
        c = *(it);
        if (c.GetSuit() == Card::TRUMPS)
        {
            stats.hasTrump = true;
            if (c.GetValue() > stats.highestTrumpValue)
            {
                stats.highestTrumpValue = c.GetValue();
            }
        }
        else
        {
            if (c.GetSuit() == suit)
            {
                stats.hasSuit = true;
            }
        }
    }

    // Card type requested is a trump
    if (suit == Card::TRUMPS)
    {
        ret = TestPlayTrump(card, stats);
    }
    // Card type requested is a standard card
    else
    {
        // The card is the required suit
        if (card.GetSuit() == suit)
        {
            ret = true;
        }
        else if (stats.hasSuit == true)
        {
            // not the required card, but he has the suit in hands
            // he must play the required suit
            ret = false;
        }
        else
        {
            // We are here if the player has not the requested suit
            ret = TestPlayTrump(card, stats);
        }
    }
    return ret;
}
/*****************************************************************************/
/**
 * @brief Player::TestPlayTrump
 *
 * This method test if the player can play a trump
 *
 * @param cVerif
 * @param hasTrump
 * @param maxPreviousTrump
 * @return
 */
bool Player::TestPlayTrump(const Card &card, const PlayerStats &stats)
{
    bool ret = false;

    // He must play a trump if he has some, higher than the highest previous played trump,
    // or any other cards in other case
    if (card.GetSuit() == Card::TRUMPS)
    {
        // He may have to play a higher trump
        if (stats.previousTrump == true)
        {
            if (card.GetValue() > stats.maxPreviousTrump)
            {
                // higher card, ok!
                ret = true;
            }
            else
            {
                // does he have a higher trump in hands?
                if (stats.highestTrumpValue > stats.maxPreviousTrump)
                {
                    ret = false; // yes, he must play it
                }
                else
                {
                    ret = true;
                }
            }
        }
        else
        {
            // No any previous trump played, so he can play any value
            ret = true;
        }
    }
    else
    {
        // Does he have a trump?
        if (stats.hasTrump == true)
        {
            // If he has only one trump and this trump is the fool, then he can play any card
            if (stats.highestTrumpValue == 0)
            {
                ret = true;
            }
            else
            {
                ret = false; // he must play a trump
            }
        }
        else
        {
            ret = true; // he can play any card
        }
    }

    return ret;
}
/*****************************************************************************/
bool Player::TestHandle(const Deck &handle)
{
    bool ret = true;
    Deck::Statistics stats;

    // Check if the handle size is correct
    if (Tarot::GetHandleType(handle.Size()) == Tarot::NO_HANDLE)
    {
        ret = false;
    }

    // Test if the handle contains only trumps
    stats.Reset();
    handle.AnalyzeTrumps(stats);

    if (handle.Size() != stats.trumps)
    {
        ret = false;
    }

    // Test if the player has all the cards of the declared handle
    for (Deck::ConstIterator it = handle.Begin(); it != handle.End(); ++it)
    {
        Card c = (*it);
        if (!HasCard(c))
        {
            ret = false;
        }
    }

    // If the fool is shown, then it indicates that there is no more any trumps in the player's hand
    stats.Reset();
    AnalyzeTrumps(stats);
    if ((handle.HasFool() == true) && (stats.trumps > handle.Size()))
    {
        ret = false;
    }
    return ret;
}
/*****************************************************************************/
bool Player::TestDiscard(const Deck &discard, const Deck &dog, std::uint8_t numberOfPlayers)
{
    bool valid = true;

    if (discard.Size() == Tarot::NumberOfDogCards(numberOfPlayers))
    {
        for (Deck::ConstIterator i = discard.Begin(); i != discard.End(); ++i)
        {
            Card c = (*i);

            // Look if the card belongs to the dog or the player's deck
            if (HasCard(c) || dog.HasCard(c))
            {
                // Look the card value against the Tarot rules
                if ((c.GetSuit() == Card::TRUMPS) ||
                        ((c.GetSuit() != Card::TRUMPS) && (c.GetValue() == 14U)))
                {
                    valid = false;
                }

                // Look if this card is unique
                if (discard.Count(c) != 1U)
                {
                    valid = false;
                }
            }
            else
            {
                valid = false;
            }
        }
    }
    else
    {
        valid = false;
    }

    return valid;
}
/*****************************************************************************/
void Player::RemoveDuplicates(const Deck &cards)
{
    // remove cards in the player's deck that are similar to the discard
    for (Deck::ConstIterator it = cards.Begin(); it != cards.End(); ++it)
    {
        Card c = (*it);
        if (HasCard(c))
        {
            Remove(c);
        }
    }
}
/*****************************************************************************/
bool Player::IsFree()
{
    if (mUuid == 0U)
    {
        return true;
    }
    else
    {
        return false;
    }
}

//=============================================================================
// End of file Player.cpp
//=============================================================================

File: ai/tarotlib/player.js
/*=============================================================================
 * TarotClub - player.js
 *=============================================================================
 * Tarot deck Javascript class
 *=============================================================================
 * TarotClub ( http://www.tarotclub.fr ) - This file is part of TarotClub
 * Copyright (C) 2003-2999 - Anthony Rabine
 * anthony@tarotclub.fr
 *
 * TarotClub is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * TarotClub is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with TarotClub.  If not, see <http://www.gnu.org/licenses/>.
 *
 *=============================================================================
 */

// namespace 
this.TarotLib = this.TarotLib||{};

(function() {

/**
 * Player class
 *
 * This class is used to manage a Tarot player
 */
var Player = function(place) {
    this.ctor(place);
};

var p = Player.prototype;

// ****************************************************************************
// PUBLIC PROPERTIES
// ****************************************************************************

// ****************************************************************************
// PRIVATE PROPERTIES
// ****************************************************************************
    this.place = 0;
    this.deck = null;
    this.stats = null;

    /**
     * Table of suits, for each one, true if suit is available for that player
     */
    this.hasSuits = null;

// ****************************************************************************
// CONSTRUCTOR
// ****************************************************************************
    p.ctor = function(place)
	{
		this.hasSuits = new Array(5);
		this.place = place;
        this.deck = new TarotLib.Deck();
        this.stats = new TarotLib.Stats();

        this.initialize();
	};

// ****************************************************************************
// PUBLIC STATIC METHODS
// ****************************************************************************

	p.initialize = function()
	{
		this.hasSuits[TarotLib.Suit.SPADES] = true;
		this.hasSuits[TarotLib.Suit.HEARTS] = true;
		this.hasSuits[TarotLib.Suit.CLUBS] = true;
		this.hasSuits[TarotLib.Suit.DIAMONDS] = true;
		this.hasSuits[TarotLib.Suit.TRUMPS] = true;
	}

    p.print = function()
    {
        for (var suit in this.hasSuits) {
            if (this.hasSuits[suit] === false) {
                systemPrint("Player " + TarotLib.Place.toString(this.place) + " has a missed suit: " + suit);
            }
        }
    };

    p.setMissedColor = function(color)
    {
        this.hasSuits[color] = false;
    };

    // return true if there is at least one suit missing, including trumps
    p.hasMissedSuit = function()
    {
        var missed = false;

        for (var i=0; i<this.hasSuits.length; i++) {
            if (this.hasSuits[i] === false) {
                missed = true;
            }
        }
        return missed;
    };

    p.updateStats = function()
    {
        this.stats.update(this.deck);
    };

    /**
     * @brief Take the lowest card in the bot's deck
     * @param[in] The suit where to get the card 
     */
    p.takeLowestCard = function(suit, minValue)
    {
        var index = 100; // init with impossible index
        var startValue = 25; // init with impossible value
        for (var i=0; i<this.deck.size(); i++)
        {
            var card = this.deck.get(i);
            if (card.suit == suit)
            {
                if ((card.value < startValue) &&
                    (card.value >= minValue))
                {
                    index = i;
                }
            }
        }
        if (index != 100)
        {
            return this.deck.takeCard(index);
        }   
        return undefined;
    };

    p.playLongSuit = function()
    {
        var playedCard = undefined;

        if (this.stats.longSuits != 0)
        {
            if (this.stats.clubs >= 5)
            {
                playedCard = this.takeLowestCard(TarotLib.Suit.CLUBS, 1);
            }
            else if (this.stats.diamonds >= 5)
            {
                playedCard = this.takeLowestCard(TarotLib.Suit.DIAMONDS, 1);
            }
            else if (this.stats.spades >= 5)
            {
                playedCard = this.takeLowestCard(TarotLib.Suit.SPADES, 1);
            }
            else
            {
                playedCard = this.takeLowestCard(TarotLib.Suit.HEARTS, 1);
            }
        }
        return playedCard;
    };

    p.playLowestCard = function(forceSuit)
    {
        var playedCard = undefined;

        if (this.stats.clubs > 0)
        {
            if ((forceSuit == TarotLib.Suit.CLUBS) ||
                (forceSuit == undefined))
            {
                playedCard = this.takeLowestCard(TarotLib.Suit.CLUBS, 1);
            }
        }
        else if (this.stats.diamonds > 0)
        {
            if ((forceSuit == TarotLib.Suit.DIAMONDS) ||
                (forceSuit == undefined))
            {
                playedCard = this.takeLowestCard(TarotLib.Suit.DIAMONDS, 1);
            }
        }
        else if (this.stats.spades > 0)
        {
            if ((forceSuit == TarotLib.Suit.SPADES) ||
                (forceSuit == undefined))
            {
                playedCard = this.takeLowestCard(TarotLib.Suit.SPADES, 1);
            }
        }
        else if (this.stats.hearts > 0)
        {
            if ((forceSuit == TarotLib.Suit.HEARTS) ||
                (forceSuit == undefined))
            {
                playedCard = this.takeLowestCard(TarotLib.Suit.HEARTS, 1);
            }
        }
        else
        {
            if ((forceSuit == TarotLib.Suit.TRUMPS) ||
                (forceSuit == undefined))
            {
                // avoid to play the one of trumps!
                playedCard = this.takeLowestCard(TarotLib.Suit.TRUMPS, 2);
            }
        }

        if (playedCard == undefined)
        {
            // whatever, play the little trumps or the fool ...
            playedCard = this.takeLowestCard(TarotLib.Suit.TRUMPS, 0);
        }

        return playedCard;
    };

    p.calculateBid = function()
    {
        var total = 0;
        this.updateStats();
        

        // We start looking at bouts, each of them increase the total value of points
        if( this.stats.bigTrump === true ) {
          total += 9;
        }
        if( this.stats.fool === true ) {
          total += 7;
        }
        if( this.stats.littleTrump === true ) {
           if( this.stats.trumps === 5 ) {
             total += 5;
          } else if( (this.stats.trumps === 6) || (this.stats.trumps === 7) ) {
             total += 7;
          } else if( this.stats.trumps > 7 ) {
             total += 8;
          }
        }

        // Each atout counts two points
        // Each major atout counts one more point
        total += this.stats.trumps * 2;
        total += this.stats.majorTrumps * 2;
        total += this.stats.kings * 6;
        total += this.stats.queens * 3;
        total += this.stats.knights * 2;
        total += this.stats.jacks;
        total += this.stats.weddings;
        total += this.stats.longSuits * 5;
        total += this.stats.cuts * 5;
        total += this.stats.singletons * 3;
        total += this.stats.sequences * 4;

        // We decide on a bid depending of thresholds
        if( total <= 35 ) {
          cont = TarotLib.Contract.PASS;
        } else if( total >= 36  && total <= 50 ) {
          cont = TarotLib.Contract.TAKE;
        } else if( total >= 51  && total <= 65 ) {
          cont = TarotLib.Contract.GUARD;
        } else if( total >= 66  && total <= 75 ) {
          cont = TarotLib.Contract.GUARD_WITHOUT;
        } else {
          cont = TarotLib.Contract.GUARD_AGAINST;
        }
        return cont;
    };

TarotLib.Player = Player;
}());
// End of file
